src/shortestpaths/johnson.jl

```diff
@@ -0,0 +1,96 @@
+
+"""
+    struct JohnsonState{T, U}
+An [`AbstractPathState`](@ref) designed for Johnson shortest-paths calculations.
+"""
+struct JohnsonState{T<:Real, U<:Integer} <: AbstractPathState
+    dists::Matrix{T}
+    parents::Matrix{U}
+end
+
+
+
+@doc_str """
+johnson_shortest_paths(g, distmx=weights(g); parallel = false)
+Use the [Johnson algorithm](https://en.wikipedia.org/wiki/Johnson%27s_algorithm)
+to compute the shortest paths between all pairs of vertices in graph `g` using an
+optional distance matrix `distmx`.
+If the parameter parallel is set true, dijkstra_shortest_paths will run in parallel.
+Parallel bellman_ford_shortest_paths is currently unavailable
+Return a [`LightGraphs.JohnsonState`](@ref) with relevant
+traversal information.
+Behaviour in case of negative cycle depends on bellman_ford_shortest_paths.
+Throws NegativeCycleError() if a negative cycle is present.
+### Performance
+Complexity: O(|V|*|E|)
+Prefered over Floyd Warshall when graph is sparse.
+If the graph is undirected (and has positive weights) then use Dijkstra.
+Undirected graph with negative weights will throw negative cycle error.
+### Dependencies from LightGraphs
+bellman_ford_shortest_paths
+parallel_multisource_dijkstra_shortest_paths
+parallel_multisource_dijkstra_shortest_paths
+"""
+function johnson_shortest_paths(
+    g::AbstractGraph{U},
+    distmx::AbstractMatrix{T} = weights(g);
+    parallel::Bool = false
+) where T<:Real where U<:Integer
+
+    nvg = nv(g)
+    #Change when parallel implementation of Bellman Ford available
+    wt_transform = bellman_ford_shortest_paths(g, vertices(g), distmx).dists
+    #Some matrices are immutable
+    new_distmx = convert(SparseMatrixCSC, distmx) 
+
+    #Transform weights
+    for e in edges(g)
+    	new_distmx[src(e), dst(e)] += wt_transform[src(e)] - wt_transform[dst(e)] 
+    end
+    dists = Matrix{T}(nvg, nvg)
+    parents = Matrix{U}(nvg, nvg)
```

Comment:
so right now we're looking at 3 * O(|V|^2) memory? This seems like a lot.

If this is really the memory footprint of Johnson's, then we'll want to note this in a `### Performance` section in the docstring.

Reply:
I could reduce the memory by | E | by not creating new_distmx 
1) Get a guarantee that the matrix is muteable
2) Create a new Dijkstra function that is the same as the current Dijkstra except that distmx[a, b] is replaced by distmx[a, b]+wt_transform[b]-wt_transform[b] 

(2) Would reduce performance but I don't know if (1) is possible. (Refer to the previous comment on general purpose all_pairs_shortest_path)

Reply:
Also 2*O(|V|^2) is expected from every all pairs shortest paths algorithm since it must be return dists and parents as output.
new_distmx will take |E| overhead but that could be avoided as explained earlier.